File: benchmarks/pandas/bench_dataframe_isin_fn.py
"""Benchmark: DataFrame.isin — test membership of each element against value sets."""
import json, time
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 50

df = pd.DataFrame({
    "a": [i % 20 for i in range(SIZE)],
    "b": [["x", "y", "z", "w"][i % 4] for i in range(SIZE)],
    "c": [i % 10 for i in range(SIZE)],
})

global_values = [0, 1, 2, "x", "y"]
col_values = {"a": [0, 1, 2, 3, 4], "b": ["x", "y"], "c": [0, 5]}

for _ in range(WARMUP):
    df.isin(global_values)
    df.isin(col_values)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    df.isin(global_values)
    df.isin(col_values)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function": "dataframe_isin_fn", "mean_ms": round(total_ms / ITERATIONS, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_dataframe_isin_fn.ts
/**
 * Benchmark: dataFrameIsin — test membership of each element in a DataFrame against value sets.
 * Outputs JSON: {"function": "dataframe_isin_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { dataFrameIsin, DataFrame } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 50;

const df = new DataFrame({
  a: Array.from({ length: SIZE }, (_, i) => i % 20),
  b: Array.from({ length: SIZE }, (_, i) => ["x", "y", "z", "w"][i % 4]),
  c: Array.from({ length: SIZE }, (_, i) => i % 10),
});

// Global isin — check all columns
const globalValues = [0, 1, 2, "x", "y"];
// Per-column isin dict
const colValues = { a: [0, 1, 2, 3, 4], b: ["x", "y"], c: [0, 5] };

for (let i = 0; i < WARMUP; i++) {
  dataFrameIsin(df, globalValues);
  dataFrameIsin(df, colValues);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  dataFrameIsin(df, globalValues);
  dataFrameIsin(df, colValues);
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "dataframe_isin_fn",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
